src/t_hash.c

```diff
@@ -1111,6 +1526,279 @@ static void hrandfieldReplyWithListpack(writePreparedClient *wpc, unsigned int c
+            if (basetime != 0 || unit == UNIT_SECONDS) {
+                robj *when_obj = createStringObjectFromLongLong(when);
+                rewriteClientCommandArgument(c, 2, when_obj);
+                decrRefCount(when_obj);
+            }
+            notifyKeyspaceEvent(NOTIFY_HASH, "hexpire", c->argv[1], c->db->id);
+        }
+        server.dirty += (expired + updated); // in case there was a change increment the dirty
+        signalModifiedKey(c, c->db, obj);
+        /* Delete the object in case it was left empty */
+        if (hashTypeLength(obj) == 0) {
+            dbDelete(c->db, c->argv[1]);
+            notifyKeyspaceEvent(NOTIFY_GENERIC, "del", c->argv[1], c->db->id);
+        }
+    }
+}
+
+void hexpireCommand(client *c) {
+    hexpireGenericCommand(c, commandTimeSnapshot(), UNIT_SECONDS);
+}
+
+void hexpireAtCommand(client *c) {
+    hexpireGenericCommand(c, 0, UNIT_SECONDS);
+}
+
+void hpexpireCommand(client *c) {
+    hexpireGenericCommand(c, commandTimeSnapshot(), UNIT_MILLISECONDS);
+}
+
+void hpexpireAtCommand(client *c) {
+    hexpireGenericCommand(c, 0, UNIT_MILLISECONDS);
+}
+
+/* High-Level Algorithm of HPERSIST Command:
+ *
+ * - Expects a key and a list of hash fields whose expiration metadata should be removed.
+ * - Validates that the number of provided fields matches the declared count.
+ *
+ * - For each specified field attempts to remove any existing expiration.
+ * - Replies to the client  with an array of integers, each representing the result of persistence for one field:
+ *   - 1 if the expiration was set.
+ *   - -1 if the field exists, but has no expiraiton time set.
+ *   - -2 if the field does not exist or the hash is empty.
+ *
+ * - If any expirations were removed:
+ *   - Marks the key as modified (for replication/AOF consistency).
+ *   - Emits a "hpersist" keyspace notification.
+ *
+ * Keyspace Notifications (if enabled):
+ * - "hpersist" — emitted once if any field had its expiration removed. */
+void hpersistCommand(client *c) {
+    int fields_index = 4, result = 0, changes = 0;
+    long long num_fields = 0;
+
+    if (getLongLongFromObjectOrReply(c, c->argv[fields_index - 1], &num_fields, NULL) != C_OK) return;
+
+    /* Check that the parsed fields number matches the real provided number of fields */
+    if (!num_fields || num_fields != (c->argc - fields_index)) {
+        addReplyErrorObject(c, shared.syntaxerr);
+        return;
+    }
+
+    /* From this point we would return array reply */
+    addReplyArrayLen(c, num_fields);
+
+    robj *hash = lookupKeyWrite(c->db, c->argv[1]);
+    if (checkType(c, hash, OBJ_HASH))
+        return;
+
+    for (int i = 0; i < num_fields; i++, fields_index++) {
+        result = hashTypePersist(hash, c->argv[fields_index]->ptr);
+        server.dirty += (result > 0 ? 1 : 0); // in case there was a change increment the dirty
+        changes += (result > 0 ? 1 : 0);
+        addReplyLongLong(c, result);
```

Comment:
Here, my preference way to write as:
if (result > 0) {
    server.dirty++;
    changes++;
}
Then we do not need to do the calculation every time.  But your solution is correct as well. You do not need to change it.

Reply:
sure. I think you point out a clearer style. will change it